### PR TITLE
Fix Email Verification

### LINKED ISSUE #1613: Email Verification Problems
Currently email verification works in below logic;

1. User registers
2. Pending mail is dispatched
3. Either automatically or by admin/staff user gets accepted
4. Accepted mail is dispatched
5. User tries to log in
6. Gets a warning about email verification

Here the problem starts (base of this issue)

7. Email verification mail is not dispatched automatically, user needs to click "send a new one" link
8. Verification mail is dispatched
9. User clicks the link (or the button)
10. Gets 403 Not Authorized error (if he/she is using different browsers for this action, imagine like login is done from Opera, but the mail is opened from phone or Chrome etc.)

**Version**
beta-5 and up

**To Reproduce**
1. Register as a new user to your own VA with a different email
2. Follow the steps as usual and use a different browser to click the verification link
3. See the error

**Expected behavior**
Technically speaking, I think the verification should happen just after registering the VA, not after being accepted. Admins should be sure that the pending application is valid and ready to be accepted.

So in theory; 
Register > Verification Mail > E-Mail Verification > Pending Mail > Accept or Reject (auto or by admin/staff) > Result Mail
should be the order of things.

And users should not be forced to click "send verification link" after waiting it to arrive, this process should be automated fully. Something like `We have send your verification link at 18:59 to your ...@.... address, if you do not receive it, please check spam folder and request a new one after ... minutes" would be nice in that blade.

**Additional context**
Email verification should be tied to an admin setting (like auto accept), people may want to disable it at all. This should not be a forced feature.

Also we may consider adding a second field to double check the mail entered during registration, like the password... Password + Confirm Password , Email + Confirm Email fields to be sure it is entered properly twice.

### PR DESCRIPTION
**Pull request by arthurpar06**
Close #1613

### CODE REVIEW

**Comment by arthurpar06** [comment]
I still need to make some changes before merge, currently if the usre is pending he cannot verify his email because of the login page which displays him in the pending view

File: app/Services/UserService.php [hunk 1]
```diff
@@ -115,7 +115,7 @@ public function createUser(array $attrs, array $roles = []): User
         $this->calculatePilotRank($user);
         $user->refresh();
 
-        event(new UserRegistered($user));
+        event(new Registered($user));
```
Comment (nabeelio) [inline app/Services/UserService.php:118]: Is this a Laravel built-in event now?

Comment (arthurpar06) [reply app/Services/UserService.php:118]: Yeah this is the build-in event to send the email verification link

https://laravel.com/docs/10.x/verification#model-preparation

**Comment by arthurpar06** [comment]
The problem is that to verify his email the user has to be logged in but he cann't login because if he is pending his login attempt is automatically refused. 

@nabeelio what do you think about instead blocking pages that require a connection via middleware if the user is pending? Because I don't really know how to allow him to verify his email without having been accepted otherwise

**Comment by nabeelio** [comment]
The verification pages should be public and not behind auth. Those endpoints can be moved into the top group which doesn't have it.

**Comment by arthurpar06** [comment]
well the thing is that laravel can't know who he's verifying if we're not logged in. This is what causes the 503 error we talk about in the issue

File: app/Http/Controllers/Admin/UserController.php [hunk 2]
```diff
@@ -323,6 +323,50 @@ public function regen_apikey(int $id, Request $request): RedirectResponse
         return redirect(route('admin.users.edit', [$id]));
     }
 
+    public function verify_email(int $id, Request $request): RedirectResponse
```
Comment (nabeelio) [inline app/Http/Controllers/Admin/UserController.php:326]: Are sure these should be in the admin route only? Seems like they should also be in the frontend user controller

Comment (nabeelio) [reply app/Http/Controllers/Admin/UserController.php:327]: Nevermind, I see what these are for

Comment (arthurpar06) [reply app/Http/Controllers/Admin/UserController.php:327]: These methods are just to allow an admin to reset the verification of an email (this way the user has to verify their email again) or to verify a user's email manually without them having to do i. These are functionalities which have been requested in the issue

**Comment by arthurpar06** [comment]
ok well now all that remains is to find a way to allow someone to check their email even if he is pending

**Comment by nabeelio** [comment]
When the registration email comes, what's the link that it sends? Do you have an example?

**Comment by arthurpar06** [comment]
http://localhost/email/verify/9/dd4a9222a7d85c79de4d82f2dd7f7c9665b0c4f4?expires=1697489889&signature=59f479306dcabca9bb5bc5af4f2ce74dc03d5b34c435835a8999d0dd28e7a490

**Comment by nabeelio** [comment]
If you comment out this line `app/Http/Controllers/Auth/VerificationController.php:36`?

**Comment by arthurpar06** [comment]
I think it calls (this)[https://github.com/laravel/ui/blob/4.x/auth-backend/VerifiesEmails.php#L35] method, which needs `$request->user()` (the logged in user)

**Comment by arthurpar06** [comment]
> If you comment out this line `app/Http/Controllers/Auth/VerificationController.php:36`? And set the redirect to login

https://flareapp.io/share/VmeWLZn7

**Comment by FatihKoz** [comment]
If this is how they built email verification, then this is really strange... Why do we need the auth and user while we are having two auto generated strings, if we have the signature (and can save it somewhere) we can easily find which user it is from that string (think it like an api key and api authentication)

**Comment by nabeelio** [comment]
I think we are missing something

**Comment by arthurpar06** [comment]
Unfortunately I'm not sure. Laravel doesn't even store the tokens that they put in the url before sending it by email, they rely everything on being logged in. I searched a little in their discord and I also found posts where moderators confirm that you have to be logged in to be able to verify your email

**Comment by arthurpar06** [comment]
This is the `.env` config for mailpit

```
MAIL_MAILER=smtp
MAIL_HOST=mailpit
MAIL_PORT=1025
MAIL_USERNAME=null
MAIL_PASSWORD=null
MAIL_ENCRYPTION=null
MAIL_FROM_ADDRESS=hello@world.com
MAIL_FROM_NAME="${APP_NAME}"
```

Then just go on `localhost:8025` and enjoy

**Comment by nabeelio** [comment]
And yes, the login to verify makes sense, it's just another level of validation that it's a real user account. And I'm using this PR branch

**Comment by arthurpar06** [comment]
Do you have auto accept new pilot enabled ?

**Comment by nabeelio** [comment]
Ah - got it. Ok, let me try with that

**Comment by nabeelio** [comment]
Whoa, styleCI is automatically applying fixes now? Hmm

**Comment by arthurpar06** [comment]
> Whoa, styleCI is automatically applying fixes now? Hmm

No, I just configured it like that on my fork because as I work quite a bit on filament in parallel and there is a lot of code it's nice not to worry and let it do its thing. But it's only on my fork/my PRs not on the main repo

**Comment by nabeelio** [comment]
Ah, how'd you configure that for PRs?

**Comment by arthurpar06** [comment]
I think it's because the PR is hosted in a branch of my fork

**Comment by arthurpar06** [comment]
Ok I still have a little trouble understanding what this piece of code allows.
```
             $user->state = UserState::ACTIVE;
             $this->userSvc->changeUserState($user, UserState::PENDING);
```

When verifying an email we don't want to change the state, we want to keep it as is. What was triggered before during a UserRegistered event will now be triggered during a Verified event. So the state will be changed in the listener if necessary, why touch it now ??

**Comment by nabeelio** [comment]
Ah you're right - let me remove those 2 lines

**Comment by arthurpar06** [comment]
we can also remove the UserService from the constructor since we don't use it anymore and I think we will be good

**Comment by arthurpar06** [comment]
I'm going to go to sleep so I haven't tested everything in detail but it seems good to me too yeah. I think we can merge.

**Pull request merged**